server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -1276,6 +1276,15 @@ private void ensureMutable() {
         }
     }
 
+    public boolean hasAllocationFailures() {
+        return unassignedShards.stream().anyMatch((shardRouting -> {
+            if (shardRouting.unassignedInfo() == null) {
+                return false;
+            }
+            return shardRouting.unassignedInfo().getNumFailedAllocations() > 0;
```

Comment:
According to https://github.com/elastic/elasticsearch/blob/6b419c13d318a14d33fdfdc9f209654db4035bcc/server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java#L108-L135, the initializing shard might have `unassignedInfo` as well. 
I do not think we need to reset error if it is attempting to start, only if it is stuck in `UNASSIGNED` state.

Reply:
My understanding that once shard is in initializing state we remove it from `unassignedShards`, isn't it? 

If yes then `unassignedShards.stream()` should have only unassigned. But this only a decision to trigger reroute with reset. The reroute will call`resetFailedCounter` and iterate over` unassignedIterator` when resets counters.